src/comparator.rs

```diff
@@ -1,30 +1,482 @@
+                exti.is_pending($Event, edge)
+            }
+
+            fn unpend(&self, exti: &EXTI) {
+                exti.unpend($Event);
+            }
+        }
+    };
+}
+
+comparator_ext!(COMP1, Comparator<COMP1>, ExtiEvent::COMP1);
+comparator_ext!(COMP2, Comparator<COMP2>, ExtiEvent::COMP2);
+
+/// Uses two comparators to implement a window comparator.
+/// See Figure 69 in RM0444 Rev 5.
+pub struct WindowComparator<U, L> {
+    pub upper: Comparator<U>,
+    pub lower: Comparator<L>,
+}
+
+pub trait WindowComparatorExt<UC, LC> {
+    /// Uses two comparators to implement a window comparator.
+    /// See Figure 69 in RM0444 Rev 5.
+    fn init<I: PositiveInput<UC>, L: NegativeInput<LC>, U: NegativeInput<UC>>(
+        &mut self,
+        input: I,
+        lower_threshold: L,
+        upper_threshold: U,
+        config: Config,
+    );
+
+    /// Returns `true` if the input is between the lower and upper thresholds
+    fn output(&self) -> bool;
+    fn output_pin<P: OutputPin<UC>>(&self, pin: P);
+    /// Returns `true` if the input is above the lower threshold
+    fn above_lower(&self) -> bool;
+    fn enable(&self);
+    fn disable(&self);
+    fn listen(&self, edge: SignalEdge, exti: &mut EXTI);
+    fn unlisten(&self, exti: &mut EXTI);
+    fn is_pending(&self, edge: SignalEdge, exti: &EXTI) -> bool;
+    fn unpend(&self, exti: &EXTI);
+}
+
+macro_rules! window_comparator {
+    ($UPPER:ident, $LOWER:ident, $LOTHR:expr) => {
+        impl WindowComparatorExt<$UPPER, $LOWER> for WindowComparator<$UPPER, $LOWER> {
+            fn init<
+                I: PositiveInput<$UPPER>,
+                L: NegativeInput<$LOWER>,
+                U: NegativeInput<$UPPER>,
+            >(
+                &mut self,
+                input: I,
+                lower_threshold: L,
+                upper_threshold: U,
+                config: Config,
+            ) {
+                let mut configu = config.clone();
+                configu.output_xor = true;
+                self.upper.init(input, upper_threshold, configu);
+
+                let mut configl = config;
+                configl.output_xor = false;
+                self.lower.init($LOTHR, lower_threshold, configl);
+            }
+
+            fn output(&self) -> bool {
+                self.upper.output()
+            }
+
+            fn output_pin<P: OutputPin<$UPPER>>(&self, pin: P) {
+                self.upper.output_pin(pin)
+            }
+
+            fn above_lower(&self) -> bool {
+                self.lower.output()
+            }
+
+            fn enable(&self) {
+                self.upper.enable();
+                self.lower.enable();
+            }
+
+            fn disable(&self) {
+                self.upper.disable();
+                self.lower.disable();
+            }
+
+            // TODO: Does the COMP need to be disabled when this is enabled?
+            // TODO: Do EXTI functions need to act on both comparators (like in ST HAL)?
```

Comment:
Could you add a brief explanation what is the actual problem and what can go wrong here.

Reply:
I've looked more closely. The ST HAL init function acts on individual comparators, and you can individually enable interrupt raising on both of them when in window mode. The default IRQHandler then checks for and clears both interrupt flags. This isn't necessary in our case since the WindowComparator constructor has control over both.